### ISSUE TITLE
static linked openssl (1.1.1i) causes Segmentation fault before main()

### ISSUE THREAD

**Issue by winter2020** [openssl/openssl#13872]
```
#include <stdio.h>
#include <openssl/evp.h>
//#include <openssl/chacha.h>

#define KEY_SIZE 32

unsigned char pt[40] = {0};
unsigned char cp[40] = {0};

void handleErrors(void)
{
    ERR_print_errors_fp(stderr);
    abort();
}

int main(int argc, char** argv) {
    unsigned char key_buf[KEY_SIZE] = {0};
    unsigned char iv[8] = {0};

    if (argc < 2) {
        fprintf(stderr, "No inputfile\n");
    }   

    FILE* pf = NULL;
    pf = fopen(argv[1], "r");
    if (!pf) {
        fprintf(stderr, "Cannot open inputfile\n");
        return 0;
    }   

    fread(key_buf, 1, KEY_SIZE, pf);
    fclose(pf);

    

    int out_len;
    EVP_CIPHER_CTX *ctx;
    ctx = EVP_CIPHER_CTX_new();
    if (!ctx)
        handleErrors();

    int rr = EVP_EncryptInit_ex(ctx, EVP_chacha20(), NULL, key_buf, iv);
    if (rr != 1)
        handleErrors();
    rr = EVP_EncryptUpdate(ctx, cp, &out_len, pt, 32);

    if (rr != 1)
        handleErrors();
 
    cp[0] = cp[1] + cp[2];

    return 0;
}
```
```
gcc test.c -g -o test -lssl -lcrypto -lpthread -ldl -static -L ~/bench/openssl -~/openssl/include/
```
Run and get an error:
```
Program received signal SIGSEGV, Segmentation fault.
0x0000000000000000 in ?? ()
(gdb) bt
#0  0x0000000000000000 in ?? ()
#1  0x000000000059a5d6 in __register_frame_info.part.4 ()
#2  0x000000000040160d in frame_dummy ()
#3  0x000000000059b847 in __libc_csu_init ()
#4  0x000000000059acfe in generic_start_main ()
#5  0x000000000059b345 in __libc_start_main ()
#6  0x0000000000401539 in _start ()

``

**Comment by KJTsanaktsidis** [comment]
I ran into this and dug into it probably more than is good for me.

If you disassemble frame 1 in your backtrace above, you probably get something like this:

```
...
   0x00000000005dabcd <+77>:    je     0x5dabdb <__register_frame_info+91>
   0x00000000005dabcf <+79>:    lea    0x1cc50a(%rip),%rdi        # 0x7a70e0 <object_mutex>
   0x00000000005dabd6 <+86>:    call   0x0
=> 0x00000000005dabdb <+91>:    mov    0x1cc536(%rip),%rax        # 0x7a7118 <unseen_objects>
   0x00000000005dabe2 <+98>:    mov    %rbx,0x1cc52f(%rip)        # 0x7a7118 <unseen_objects>
   0x00000000005dabe9 <+105>:   mov    %rax,0x28(%rbx)
   0x00000000005dabed <+109>:   mov    0x1cc515(%rip),%eax        # 0x7a7108 <any_objects_registered>
   0x00000000005dabf3 <+115>:   test   %eax,%eax
...
```

Obviously calling 0x0 is going to segfault, but why is the compiler putting that in there? That question lead me to this SO answer: https://stackoverflow.com/questions/46350290/what-is-the-sense-of-this-asm-code which indicated that it was probably the result of the weak pthread symbols not being linked in, and so they get set to zero.

I figured out I needed to compile my program with something like this, then it works fine.

```
gcc testprogram.c -o testprogram -static -lssl -lcrypto -ldl -Wl,--whole-archive -lpthread -Wl,--no-whole-archive
``` 

Just using `-lpthread` was NOT enough - my reading of this SO answer(https://stackoverflow.com/questions/13089166/how-to-make-gcc-link-strong-symbol-in-static-library-to-overwrite-weak-symbol) is that the static linker won't bother linking in the real pthread functions if there's already a definition of them in an earlier object. But using `--whole-archive` for pthread will force it to overwrite the weak symbols.

TL;DR: Add `-Wl,--whole-archive -lpthread -Wl,--no-whole-archive` to your compile command.

**Comment by nhorman** [comment]
Marking as inactive, to be closed at the end of the 3.4 dev cycle barring further input

**Comment by t8m** [comment]
Closing as the issue was labeled as inactive for extended period of time and there was no additional input.
Please add a comment if you have any further information to support reopening the issue.
